chain/checkpointing/sub.go

```diff
@@ -0,0 +1,506 @@
+	if err != nil {
+		return
+	}
+}
+
+func (c *CheckpointingSub) Start(ctx context.Context) {
+	topic, err := c.pubsub.Join("keygen")
+	if err != nil {
+		panic(err)
+	}
+	c.topic = topic
+
+	// and subscribe to it
+	sub, err := topic.Subscribe()
+	if err != nil {
+		panic(err)
+	}
+	c.sub = sub
+
+	c.listenCheckpointEvents(ctx)
+}
+
+func (c *CheckpointingSub) LoopHandler(ctx context.Context, h protocol.Handler, network *Network) {
+	for {
+		msg, ok := <-h.Listen()
+		if !ok {
+			network.Done()
+			// the channel was closed, indicating that the protocol is done executing.
+			fmt.Println("Should be good")
+			return
+		}
+		network.Send(ctx, msg)
+
+		for _, _ = range network.Parties() {
+			msg = network.Next(ctx)
+			h.Accept(msg)
+		}
+	}
+}
+
+func (c *CheckpointingSub) CreateCheckpoint(ctx context.Context, cp, data []byte) {
+	idsStrings := c.orderParticipantsList()
+	fmt.Println("Participants list :", idsStrings)
+	fmt.Println("Precedent tx", c.ptxid)
+	ids := c.formIDSlice(idsStrings)
+	taprootAddress := PubkeyToTapprootAddress(c.pubkey)
+
+	pubkey := c.config.PublicKey
+	if c.newconfig != nil {
+		pubkey = c.newconfig.PublicKey
+	}
+
+	pubkeyShort := GenCheckpointPublicKeyTaproot(pubkey, cp)
+	newTaprootAddress := PubkeyToTapprootAddress(pubkeyShort)
+
+	if c.ptxid == "" {
+		fmt.Println("Missing precedent txid")
+		taprootScript := GetTaprootScript(c.pubkey)
+		success := AddTaprootScriptToWallet(taprootScript)
+		if !success {
+			panic("failed to add taproot address to wallet")
+		}
+
+		ptxid, err := WalletGetTxidFromAddress(taprootAddress)
+		fmt.Println(taprootAddress)
+		if err != nil {
+			panic(err)
+		}
+		c.ptxid = ptxid
+		fmt.Println("Found precedent txid:", c.ptxid)
+	}
+
+	index := 0
+	value, scriptPubkeyBytes := GetTxOut(c.ptxid, index)
+
+	if scriptPubkeyBytes[0] != 0x51 {
+		fmt.Println("Wrong txout")
+		index = 1
+		value, scriptPubkeyBytes = GetTxOut(c.ptxid, index)
+	}
+	newValue := value - FEE
+
+	payload := "{\"jsonrpc\": \"1.0\", \"id\":\"wow\", \"method\": \"createrawtransaction\", \"params\": [[{\"txid\":\"" + c.ptxid + "\",\"vout\": " + strconv.Itoa(index) + ", \"sequence\": 4294967295}], [{\"" + newTaprootAddress + "\": \"" + fmt.Sprintf("%.2f", newValue) + "\"}, {\"data\": \"" + hex.EncodeToString(data) + "\"}]]}"
+	result := jsonRPC(payload)
+	fmt.Println(result)
+	if result == nil {
+		panic("cant create new transaction")
+	}
+
+	rawTransaction := result["result"].(string)
+
+	tx, err := hex.DecodeString(rawTransaction)
+	if err != nil {
+		panic(err)
+	}
+
+	var buf [8]byte
+	binary.LittleEndian.PutUint64(buf[:], uint64(value*100000000))
+	utxo := append(buf[:], []byte{34}...)
+	utxo = append(utxo, scriptPubkeyBytes...)
+
+	hashedTx, err := TaprootSignatureHash(tx, utxo, 0x00)
+	if err != nil {
+		panic(err)
+	}
+
+	/*
+	 * Orchestrate the signing message
+	 */
+
+	f := frost.SignTaprootWithTweak(c.config, ids, hashedTx[:], c.tweakedValue[:])
+	n := NewNetwork(ids, c.sub, c.topic)
```

Comment:
We should definitely the change the name of the `Network` struct, because without additional docs it is very misleading.

Reply:
I will add a comment. Any name suggestions ? I re-use `Network` because it is how it was called in the lib and our script.

Reply:
`NewRound` maybe? Or something like `NewState`? Let's try to find something more self-describing than `Network`  if we can. It led me to confusion the first time I read it :)

Reply:
I thought of `NewRound` too but it could be confusing at the cryptographic level because there is several rounds of communication when you do the DKG/signing... What about `NewNetInterface` ? Because it is an interface for communication between participants. They can receive and send messages through this instance.

What about `Communication` ? `NewCommunicationRound` ?